Cookbook/Proposals/Danger-Rule-Xcodeproj.md

```diff
@@ -0,0 +1,87 @@
+# Danger Rules for pbxproj
+
+* Author: Olivier Halligon
+* Review Manager: TBD
+
+## Introduction
+
+This proposal suggests to add some Danger rules checking for `xcodeproj` consistenty.
+
+It is dependent on [#110 - Integrate Danger](https://github.com/Babylonpartners/ios-playbook/pull/110) being accepted first.
+
+## Motivation
+
+`pbxproj` project format is hard to parse, which means it's easy to miss some mistakes that could happen when an Xcode project has changed during a PR.
+
+This means that some mistakes have ended up in `develop` in the past, like:
+
+* `(null)` references in `pbxproj` files
+* Files being added as part of a target membership while they shouldn't:
+  * Snapshot files
+  * xcconfig files
+  * Info.plist files
+
+Some of those mistakes, especially `(null)` references, are generally the result of merge conflicts; so their appearances during a PR could be an indicator that something else could have gone wrong during the merge of the pbxproj file. It's generally a good incentive to double-check that there was no other side effect from the pbxproj merge.
+
+Other mistakes like unexpected files in target membership could lead to files ending up in the final bundle uploaded to the AppStore, and in addition to being useless in the final `ipa`, bloat the app size for no reason.
+
+## Proposed solution
+
+* We can use a simple "find" in every `*.pbxproj` for the string `/* (null) */`, and make Danger generate an inline comment on the found line(s) if any
```

Comment:
Will Danger automatically provide timing values for all its tasks or its something that we would need to implement ourselves? I think it's something we can be interested in monitoring to avoid that those checks are taking a very long time considering the size of our project (actually same can be applied to other tools like swiftlint @diegopetrucci @TheAdamBorek )

Reply:
Good question.
I don't think Danger provides timing info. What we could imagine doing is get the time at the beginning and end of Dangerfile and include a `message('Danger took 160ms to run its rules')` at the very end.

Not sure it's worth it though tbh, as like it's explained in the main #110 proposal, as opposed to SwiftLint — which needs to go thru every file in the project (changed or not) one by one and parse their content and even analyze the SIL representation for some rules — Danger only do some tests on the PR metadata… or grep on a very small subset of files (like one or two pbxproj which have changed in the PR), so it's way less time-consuming compared to e.g. SwiftLint.